src/dimension/dimension_trait.rs

```diff
@@ -78,6 +79,34 @@ pub trait Dimension:
     /// Next larger dimension
     type Larger: Dimension + RemoveAxis;
 
+    /// Convert index to &Self::SliceArg. Make sure that length of index
+    /// consists with Self::NDIM(if it exists).
+    ///
+    /// Panics if conversion failed.
+    #[doc(hidden)]
+    fn slice_arg_from(index: &[SliceOrIndex]) -> &Self::SliceArg ;
+
+    /// Convert &SliceInfo<AsRef<[SliceOrIndex]>, Do> to &SliceInfo<D::SliceArg, Do>.
+    /// Generate SliceArg of any dimension via this method.
+    ///
+    /// Panics if conversion failed.
+    #[doc(hidden)]
+    fn slice_info_from<T, Do>(indices: &T) -> &SliceInfo<Self::SliceArg, Do>
```

Comment:
This isn't correct, because the `Do` here is unconstrained and is not checked. Instead, we should add a method to `SliceInfo`:

```rust
impl<T: ?Sized, D> SliceInfo<T, D>
where
    T: AsRef<[SliceOrIndex]>,
    D: Dimension,
{
    pub fn for_dimensionality<E>(&self) -> Result<&SliceInfo<E::SliceArg, D>, ShapeError>
    where
        E: Dimension,
    {
        ...
    }
}
```

which makes sure that the `D` is correctly preserved.

Reply:
For what it's worth, we could do this instead to avoid the need for `Dimension::slice_arg_from`:

```rust
impl<T: ?Sized, D> SliceInfo<T, D>
where
    T: AsRef<[SliceOrIndex]>,
    D: Dimension,
{
    pub fn for_dimensionality<'a, E>(&'a self) -> &'a SliceInfo<E::SliceArg, D>
    where
        E: Dimension,
        &'a D::SliceArg: TryFrom<&'a [SliceOrIndex]>,
    {
        unimplemented!()
    }
}
```

but that would make it awkward to use in generic functions because users would have to add the `&'a D::SliceArg: TryFrom<&'a [SliceOrIndex]>` bound to their functions.

Reply:
That's right, function should be put in SliceInfo. But I think it seems difficult to move slice_arg_from out of Dimension.